lib/kafkaesque.js

```diff
@@ -114,74 +202,105 @@ module.exports = function(options) {
   };
 
 
+  var _connectToPartition = function (params, leader, cb) {
+    var topicKey = _makeTopicKey(params);
+    var leaderKey = _makeBrokerKey(leader);
+    var cacheTopic = function() {
+      _topics[topicKey] = {
+        params: params,
+        broker: _brokers[leaderKey]
+      };
+
+      _offsetPosition(_topics[topicKey], cb);
+    };
+
+    var connect = function(err) {
+      if (err) {
+        return cb(err, null);
+      } else {
+        _brokers[leaderKey].connected = true;
+        cacheTopic();
+      }
+    };
+
+    if (!_brokers[leaderKey]) {
+      _brokers[leaderKey] = api({host: leader.host, port: leader.port, maxBytes: _options.maxBytes, clientId: _options.clientId});
+      _brokers[leaderKey].tearUp(connect);
+    }
+    else {
+      if (!_brokers[leaderKey].connected) {
+        _brokers[leaderKey].tearUp(connect);
+      } else {
+        cacheTopic();
+      }
+    }
+  };
+
 
   var _initiate = function(params, cb) {
+    // console.log('init params', params)
     if (!_topics[_makeTopicKey(params)]) {
-      _determineTopicLeader(params, function(err, leader) {
-        if (err) {
-          cb(err, null);
-          return;
-        }
+      if (params.partition !== undefined) {
+        _determinePartitionLeader(params, function (err, leader) {
+          if (err) {
+            return cb(err, null);
+          }
+          // console.log('leader details', params, leader)
+          _connectToPartition(params, leader, cb);
+        });
+      } else {
+        _getTopicMetadata(params, function(err, cluster) {
+          if (err) {
+            return cb(err, null);
+          }
 
-        var leaderKey = _makeBrokerKey(leader);
-        var topicKey = _makeTopicKey(params);
+          // set _topics[_makeTopicKey(params)] = true to cause enclosing
```

Comment:
This was actually explaining what was going on with the code, so I'm inclined to leave that as is. is that okay?